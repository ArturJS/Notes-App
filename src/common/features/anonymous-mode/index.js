export { default as watchAnonymousMode } from './anonymous-mode.saga';
