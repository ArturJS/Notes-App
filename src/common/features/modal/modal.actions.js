import _ from 'lodash';
import { createActions } from 'redux-actions';

export const OPEN_MODAL = 'OPEN_MODAL';
export const CLOSE_MODAL_REQUEST = 'CLOSE_MODAL_REQUEST';
export const CLOSE_MODAL_SUCCESS = 'CLOSE_MODAL_SUCCESS';
export const CLOSE_ALL_MODALS_REQUEST = 'CLOSE_ALL_MODALS_REQUEST';
export const CLOSE_ALL_MODALS_SUCCESS = 'CLOSE_ALL_MODALS_SUCCESS';
export const {
    openModal,
    closeModalRequest,
    closeModalSuccess,
    closeAllModalsRequest,
    closeAllModalsSuccess
} = createActions({
    [OPEN_MODAL]: modalInstance => ({
        modalInstance
    }),
    [CLOSE_MODAL_REQUEST]: ({ id, reason }) => ({
        id,
        reason
    }),
    [CLOSE_MODAL_SUCCESS]: id => ({ id }),
    [CLOSE_ALL_MODALS_REQUEST]: ({ reason }) => ({ reason }),
    [CLOSE_ALL_MODALS_SUCCESS]: _.noop
});
