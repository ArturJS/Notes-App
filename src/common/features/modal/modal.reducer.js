import { handleActions } from 'redux-actions';
import produce from 'immer';
import {
    OPEN_MODAL,
    CLOSE_MODAL_REQUEST,
    CLOSE_MODAL_SUCCESS,
    CLOSE_ALL_MODALS_REQUEST,
    CLOSE_ALL_MODALS_SUCCESS
} from './modal.actions';

const initialState = [];

/* eslint-disable no-param-reassign */
const modalReducer = handleActions(
    {
        [OPEN_MODAL]: (state, { payload }) =>
            produce(state, draftState => {
                draftState.push(payload.modalInstance);
            }),
        [CLOSE_MODAL_REQUEST]: (state, { payload }) =>
            produce(state, draftState => {
                const { id, reason } = payload;
                const relatedModal = draftState.find(modal => modal.id === id);

                if (relatedModal) {
                    relatedModal.isOpen = false;

                    if (relatedModal.close) {
                        relatedModal.close(reason);
                    }
                }
            }),
        [CLOSE_MODAL_SUCCESS]: (state, { payload }) =>
            produce(state, draftState => {
                const { id } = payload;

                return draftState.filter(modal => modal.id !== id);
            }),
        [CLOSE_ALL_MODALS_REQUEST]: (state, { payload }) =>
            produce(state, draftState => {
                const { reason } = payload;

                draftState.forEach(modal => {
                    modal.isOpen = false;

                    if (modal.close) {
                        modal.close(reason);
                    }
                });
            }),
        [CLOSE_ALL_MODALS_SUCCESS]: () => []
    },
    initialState
);
/* eslint-enable no-param-reassign */

export default modalReducer;
