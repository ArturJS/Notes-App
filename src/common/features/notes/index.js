import {
    addNoteRequest,
    addNoteSuccess,
    addNoteFailure,
    updateNoteRequest,
    updateNoteSuccess,
    updateNoteFailure,
    deleteNoteRequest,
    deleteNoteSuccess,
    deleteNoteFailure,
    getAllNotesRequest,
    getAllNotesSuccess,
    getAllNotesFailure,
    changeNoteOrderRequest,
    changeNoteOrderSuccess,
    changeNoteOrderFailure,
    clearNotes
} from './notes.actions';
import * as _notesSelectors from './notes.selectors';

export const notesActions = {
    addNoteRequest,
    addNoteSuccess,
    addNoteFailure,
    updateNoteRequest,
    updateNoteSuccess,
    updateNoteFailure,
    deleteNoteRequest,
    deleteNoteSuccess,
    deleteNoteFailure,
    getAllNotesRequest,
    getAllNotesSuccess,
    getAllNotesFailure,
    changeNoteOrderRequest,
    changeNoteOrderSuccess,
    changeNoteOrderFailure,
    clearNotes
};
export const notesSelectors = _notesSelectors;
export { default as notesReducer } from './notes.reducer';
export { default as watchNotes } from './notes.saga';
