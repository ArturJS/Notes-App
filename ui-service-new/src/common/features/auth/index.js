import * as _authActions from './auth.actions';
import * as _authSelectors from './auth.selectors';

export const authActions = _authActions;
export const authSelectors = _authSelectors;
export { default as authReducer } from './auth.reducer';
export { default as watchAuth } from './auth.saga';
