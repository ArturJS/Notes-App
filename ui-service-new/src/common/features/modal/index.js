export { default as modalReducer } from './modal.reducer';
export { default as ModalDialog } from './modal-dialog';
export { default as modalProvider } from './modal.provider';
