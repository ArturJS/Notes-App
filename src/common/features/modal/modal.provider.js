import shortid from 'shortid';
import store from '~/common/store';
import { sleep } from '../../utils';
import * as modalActions from './modal.actions';

export const MODAL_TYPES = {
    error: 'ERROR_MODAL',
    confirm: 'CONFIRM_MODAL',
    info: 'INFO_MODAL',
    custom: 'CUSTOM_MODAL'
};

// it's necessary to perform exit animation in modal-dialog.jsx
const CLOSE_DELAY_MS = 10;

class Modal {
    constructor({
        id,
        title = '',
        body,
        type,
        close,
        className = '',
        shouldCloseOnOverlayClick = true,
        noBackdrop = false
    }) {
        this.id = id;
        this.title = title;
        this.body = body;
        this.type = type;
        this.close = close;
        this.className = className;
        this.shouldCloseOnOverlayClick = shouldCloseOnOverlayClick;
        this.noBackdrop = noBackdrop;
        this.isOpen = true;
    }
}

class ModalProvider {
    showConfirm({ title, body, className }) {
        const { result, close } = this._openModal({
            title,
            body,
            type: MODAL_TYPES.confirm,
            className
        });

        return {
            result,
            close
        };
    }

    showError({ title, body, className = '' }) {
        const { result, close } = this._openModal({
            title,
            body,
            type: MODAL_TYPES.error,
            className: `modal-error ${className}`,
            noBackdrop: true
        });

        return {
            result,
            close
        };
    }

    showCustom({ title, body, className }) {
        const { result, close } = this._openModal({
            title,
            body,
            type: MODAL_TYPES.custom,
            className
        });

        return {
            result,
            close
        };
    }

    async closeAll(reason) {
        store.dispatch(
            modalActions.closeAllModalsRequest({
                reason
            })
        );

        await sleep(CLOSE_DELAY_MS);

        store.dispatch(modalActions.closeAllModalsSuccess());
    }

    _openModal({ title, body, type, className, noBackdrop }) {
        let close;
        const result = new Promise(resolve => {
            close = resolve;
        });
        const id = shortid.generate();

        store.dispatch(
            modalActions.openModal(
                new Modal({
                    id,
                    title,
                    body,
                    type,
                    className,
                    close,
                    noBackdrop
                })
            )
        );

        result.then(reason => this._closeModal({ id, reason }));

        return {
            result,
            close
        };
    }

    async _closeModal({ id, reason }) {
        store.dispatch(modalActions.closeModalRequest({ id, reason }));

        await sleep(CLOSE_DELAY_MS);

        store.dispatch(modalActions.closeModalSuccess({ id }));
    }
}

export default new ModalProvider();
