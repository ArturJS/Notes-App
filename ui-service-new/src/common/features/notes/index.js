import * as _notesActions from './notes.actions';
import * as _notesSelectors from './notes.selectors';

export const notesActions = _notesActions;
export const notesSelectors = _notesSelectors;
export { default as notesReducer } from './notes.reducer';
export { default as watchNotes } from './notes.saga';
