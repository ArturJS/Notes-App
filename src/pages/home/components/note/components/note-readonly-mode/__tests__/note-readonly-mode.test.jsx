jest.mock('~/common/features/modal', () => ({
    modalProvider: {
        showConfirm: jest.fn()
    }
}));
jest.mock('../../../../file-list', () => props => (
    <div className="file-list">{JSON.stringify(props, null, '  ')}</div>
));

/* eslint-disable import/first */
import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import { modalProvider } from '~/common/features/modal';
import NoteReadonlyMode from '../note-readonly-mode';
/* eslint-enable import/first */

const flushPromises = () => new Promise(resolve => setImmediate(resolve));

describe('<NoteReadonlyMode />', () => {
    it('renders correctly', () => {
        const note = {
            id: 1,
            title: 'Note test title',
            description: 'Note test description',
            files: [
                {
                    id: 12,
                    downloadPath: '/test/download/path',
                    filename: 'file name'
                }
            ],
            trackId: '1',
            meta: {
                transactionId: 1
            }
        };
        const tree = renderer
            .create(
                <NoteReadonlyMode
                    note={note}
                    onRemove={() => {}}
                    onEdit={() => {}}
                />
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it("wraps url's in description correctly", () => {
        const note = {
            id: 1,
            title: 'Note test title',
            description: 'http://google.com and https://ya.ru/',
            files: [],
            trackId: '1'
        };
        const wrapper = mount(
            <NoteReadonlyMode
                note={note}
                onRemove={() => {}}
                onEdit={() => {}}
            />
        );

        expect(wrapper.find('.note-description').html()).toBe(
            [
                '<div class="note-description">',
                '<a ',
                'href="http://google.com" ',
                'class="note-link" ',
                'target="_blank" ',
                'rel="nofollow noopener">http://google.com</a>',
                ' and ',
                '<a ',
                'href="https://ya.ru/" ',
                'class="note-link" ',
                'target="_blank" ',
                'rel="nofollow noopener">https://ya.ru/</a>',
                '</div>'
            ].join('')
        );
    });

    describe('test isLoading state', () => {
        it('should have `isLoading` class when note has `meta.transactionId`', () => {
            const note = {
                id: 1,
                title: 'Note test title',
                description: 'Note test description',
                files: [
                    {
                        id: 12,
                        downloadPath: '/test/download/path',
                        filename: 'file name'
                    }
                ],
                trackId: '1',
                meta: {
                    transactionId: 1
                }
            };
            const wrapper = mount(
                <NoteReadonlyMode
                    note={note}
                    onRemove={() => {}}
                    onEdit={() => {}}
                />
            );

            expect(wrapper.render().hasClass('isLoading')).toBe(true);
        });

        it('should NOT have `isLoading` class when note do NOT have `meta.transactionId`', () => {
            const note = {
                id: 1,
                title: 'Note test title',
                description: 'Note test description',
                files: [
                    {
                        id: 12,
                        downloadPath: '/test/download/path',
                        filename: 'file name'
                    }
                ],
                trackId: '1'
            };
            const wrapper = mount(
                <NoteReadonlyMode
                    note={note}
                    onRemove={() => {}}
                    onEdit={() => {}}
                />
            );

            expect(wrapper.render().hasClass('isLoading')).toBe(false);
        });
    });

    describe('`onEdit` prop', () => {
        describe('it should be called after ', () => {
            it('clicked by .icon-left', () => {
                const note = {
                    id: 1,
                    title: 'Note test title',
                    description: 'Note test description',
                    files: [
                        {
                            id: 12,
                            downloadPath: '/test/download/path',
                            filename: 'file name'
                        }
                    ],
                    trackId: '1'
                };
                const onEdit = jest.fn();
                const wrapper = mount(
                    <NoteReadonlyMode
                        note={note}
                        onRemove={() => {}}
                        onEdit={onEdit}
                    />
                );

                wrapper.find('.icon-left').simulate('click');

                expect(onEdit).toHaveBeenCalledWith();
            });

            it('keyPressed by .icon-left', () => {
                const note = {
                    id: 1,
                    title: 'Note test title',
                    description: 'Note test description',
                    files: [
                        {
                            id: 12,
                            downloadPath: '/test/download/path',
                            filename: 'file name'
                        }
                    ],
                    trackId: '1'
                };
                const onEdit = jest.fn();
                const wrapper = mount(
                    <NoteReadonlyMode
                        note={note}
                        onRemove={() => {}}
                        onEdit={onEdit}
                    />
                );

                wrapper.find('.icon-left').simulate('keyPress');

                expect(onEdit).toHaveBeenCalledWith();
            });
        });

        describe('it should NOT be called (if isLoading) after ', () => {
            it('clicked by .icon-left', () => {
                const note = {
                    id: 1,
                    title: 'Note test title',
                    description: 'Note test description',
                    files: [
                        {
                            id: 12,
                            downloadPath: '/test/download/path',
                            filename: 'file name'
                        }
                    ],
                    trackId: '1',
                    meta: {
                        transactionId: 1
                    }
                };
                const onEdit = jest.fn();
                const wrapper = mount(
                    <NoteReadonlyMode
                        note={note}
                        onRemove={() => {}}
                        onEdit={onEdit}
                    />
                );

                wrapper.find('.icon-left').simulate('click');

                expect(onEdit).not.toHaveBeenCalledWith();
            });

            it('keyPressed by .icon-left', () => {
                const note = {
                    id: 1,
                    title: 'Note test title',
                    description: 'Note test description',
                    files: [
                        {
                            id: 12,
                            downloadPath: '/test/download/path',
                            filename: 'file name'
                        }
                    ],
                    trackId: '1',
                    meta: {
                        transactionId: 1
                    }
                };
                const onEdit = jest.fn();
                const wrapper = mount(
                    <NoteReadonlyMode
                        note={note}
                        onRemove={() => {}}
                        onEdit={onEdit}
                    />
                );

                wrapper.find('.icon-left').simulate('keyPress');

                expect(onEdit).not.toHaveBeenCalledWith();
            });
        });
    });

    describe('`onRemove` prop', () => {
        describe('it should be called after ', () => {
            it('clicked by .icon-right and modalProvider.showConfirm(...) confirms action', async () => {
                const note = {
                    id: 132,
                    title: 'Note test title',
                    description: 'Note test description',
                    files: [
                        {
                            id: 12,
                            downloadPath: '/test/download/path',
                            filename: 'file name'
                        }
                    ],
                    trackId: '1'
                };
                const onRemove = jest.fn();

                modalProvider.showConfirm.mockImplementation(() => ({
                    result: Promise.resolve(true)
                }));

                const wrapper = mount(
                    <NoteReadonlyMode
                        note={note}
                        onRemove={onRemove}
                        onEdit={() => {}}
                    />
                );

                wrapper.find('.icon-right').simulate('click');

                await flushPromises();

                expect(onRemove).toHaveBeenCalledWith(note.id);
            });

            it('keyPressed by .icon-right and modalProvider.showConfirm(...) confirms action', async () => {
                const note = {
                    id: 132,
                    title: 'Note test title',
                    description: 'Note test description',
                    files: [
                        {
                            id: 12,
                            downloadPath: '/test/download/path',
                            filename: 'file name'
                        }
                    ],
                    trackId: '1'
                };
                const onRemove = jest.fn();

                modalProvider.showConfirm.mockImplementation(() => ({
                    result: Promise.resolve(true)
                }));

                const wrapper = mount(
                    <NoteReadonlyMode
                        note={note}
                        onRemove={onRemove}
                        onEdit={() => {}}
                    />
                );

                wrapper.find('.icon-right').simulate('keyPress');

                await flushPromises();

                expect(onRemove).toHaveBeenCalledWith(note.id);
            });
        });

        describe('it should NOT be called after ', () => {
            describe('when isLoading and after ', () => {
                it('clicked by .icon-right and modalProvider.showConfirm(...) confirms action', async () => {
                    const note = {
                        id: 132,
                        title: 'Note test title',
                        description: 'Note test description',
                        files: [
                            {
                                id: 12,
                                downloadPath: '/test/download/path',
                                filename: 'file name'
                            }
                        ],
                        trackId: '1',
                        meta: {
                            transactionId: 1
                        }
                    };
                    const onRemove = jest.fn();

                    modalProvider.showConfirm.mockImplementation(() => ({
                        result: Promise.resolve(true)
                    }));

                    const wrapper = mount(
                        <NoteReadonlyMode
                            note={note}
                            onRemove={onRemove}
                            onEdit={() => {}}
                        />
                    );

                    wrapper.find('.icon-right').simulate('click');

                    await flushPromises();

                    expect(onRemove).not.toHaveBeenCalledWith(note.id);
                });

                it('keyPressed by .icon-right and modalProvider.showConfirm(...) confirms action', async () => {
                    const note = {
                        id: 132,
                        title: 'Note test title',
                        description: 'Note test description',
                        files: [
                            {
                                id: 12,
                                downloadPath: '/test/download/path',
                                filename: 'file name'
                            }
                        ],
                        trackId: '1',
                        meta: {
                            transactionId: 1
                        }
                    };
                    const onRemove = jest.fn();

                    modalProvider.showConfirm.mockImplementation(() => ({
                        result: Promise.resolve(true)
                    }));

                    const wrapper = mount(
                        <NoteReadonlyMode
                            note={note}
                            onRemove={onRemove}
                            onEdit={() => {}}
                        />
                    );

                    wrapper.find('.icon-right').simulate('keyPress');

                    await flushPromises();

                    expect(onRemove).not.toHaveBeenCalledWith(note.id);
                });
            });

            describe('when NOT isLoading and after ', () => {
                it('clicked by .icon-right and modalProvider.showConfirm(...) NOT confirms action', async () => {
                    const note = {
                        id: 132,
                        title: 'Note test title',
                        description: 'Note test description',
                        files: [
                            {
                                id: 12,
                                downloadPath: '/test/download/path',
                                filename: 'file name'
                            }
                        ],
                        trackId: '1'
                    };
                    const onRemove = jest.fn();

                    modalProvider.showConfirm.mockImplementation(() => ({
                        result: Promise.resolve(false)
                    }));

                    const wrapper = mount(
                        <NoteReadonlyMode
                            note={note}
                            onRemove={onRemove}
                            onEdit={() => {}}
                        />
                    );

                    wrapper.find('.icon-right').simulate('click');

                    await flushPromises();

                    expect(onRemove).not.toHaveBeenCalled();
                });

                it('keyPressed by .icon-right and modalProvider.showConfirm(...) NOT confirms action', async () => {
                    const note = {
                        id: 132,
                        title: 'Note test title',
                        description: 'Note test description',
                        files: [
                            {
                                id: 12,
                                downloadPath: '/test/download/path',
                                filename: 'file name'
                            }
                        ],
                        trackId: '1'
                    };
                    const onRemove = jest.fn();

                    modalProvider.showConfirm.mockImplementation(() => ({
                        result: Promise.resolve(false)
                    }));

                    const wrapper = mount(
                        <NoteReadonlyMode
                            note={note}
                            onRemove={onRemove}
                            onEdit={() => {}}
                        />
                    );

                    wrapper.find('.icon-right').simulate('keyPress');

                    await flushPromises();

                    expect(onRemove).not.toHaveBeenCalled();
                });
            });
        });
    });
});
