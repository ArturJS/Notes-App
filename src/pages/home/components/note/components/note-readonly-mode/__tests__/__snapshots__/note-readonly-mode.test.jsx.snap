// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoteReadonlyMode /> renders correctly 1`] = `
<div
  className="note isLoading"
>
  <div>
    <i
      className="icon icon-left icon-pencil"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    />
    <i
      className="icon icon-right icon-bin"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    />
  </div>
  <div
    className="note-title"
  >
    Note test title
  </div>
  <div
    className="note-description"
    dangerouslySetInnerHTML={
      Object {
        "__html": "Note test description",
      }
    }
  />
  <div
    className="file-list"
  >
    {
  "files": [
    {
      "id": 12,
      "downloadPath": "/test/download/path",
      "filename": "file name"
    }
  ]
}
  </div>
</div>
`;
